# **Data Types and Common Functions**
In this section, we are going to look at different data types that SyumPy provides. We are also going to check out several new constants and functions in SymPy.

### **Functions and attributes in this section:**
 - **sp.Integer()**   - Provides symbolic integers in SymPy.
 - **sp.Rational()**  - Provides symbolic fractions in SymPy.
 - **sp.pi**          - Symbolic version of π.
 - **sp.E**           - Symbolic version of e.
 - **sp.oo**          - Symbolic representation of ∞. (Double oo) 
 - **sp.I**           - Symbolic representation of the imaginary unit *i* (denoted by *j* in engineering literature and   the math package in Python). 
 - **sp.exp()**         - The exponential function. 
 - **sp.log()**         - The natural logarith. 
 - **sp.sin()**         - The Trignometric function.  
 - **sp.cos()**         - The Trignometric function. 
 - **sp.tan()**         - The Trignometric function.   

In [1]:
# import SymPy Library
import sympy as sp

### **Why Data Types in SymPy Matter!**

In [2]:
# Defining a symbol x
x = sp.Symbol("x")
x

x

In [3]:
expr = x**(1/3)
expr

x**0.333333333333333

 - Firstly, the expression 1/3 is evaluated to be the floating-point number 0.333333......
 - Secondly, the SymPy x is raised tp the power 0.333333... to form x^0.333333333333333
Involving nummeric into our symbolic mathematical makes it difficult to recognize terms: 

In [4]:
# This should be exactly x^(1/15)
expr ** (1/5)

x**0.0666666666666667

To avoid this, we will introduce the SymPy's data types **sp.Integer()** and **sp.Rational()**. 

In [5]:
# SymPy's interger type 
expr = x**(sp.Integer(1)/sp.Integer(3))
expr

x**(1/3)

In [6]:
# This should be exactly x^(1/15)
expr = x ** (sp.Integer(1)/sp.Integer(5))
expr

x**(1/5)

In [7]:
# Even more simple
expr ** (sp.Integer(1)/5)
expr

x**(1/5)

Why? Because the rule is that when there is an operation between a SymPy object and a basic Python object, then the result is converted into a Sympy object. 

In [8]:
# This automatically becomes a SymPy rational object
type(sp.Integer(1)/5)

sympy.core.numbers.Rational

### **SymPy's Rational Type**
For specifying rational numbers we can also directly use the **sp.Rational()** constructor. 

In [9]:
# SymPy's rational object
rational= sp.Rational(1,3)
rational

1/3

Hence we could also have written the following:

In [10]:
expr = (x**sp.Rational(1, 3))  ** sp.Rational(1, 5)
expr 

x**(1/15)

#### **Important Constants**

In [11]:
# The constant pi
# sp.pi
# 3 * sp.pi
15 * sp.pi
# so on... 

15*pi

In [12]:
# The Constant e 
sp.E

E

In [13]:
# Add constant pi and e 
sp.E + 15 * sp.pi

E + 15*pi

In [14]:
# Infinity
sp.oo 

oo

In [15]:
# What, If we add something to infinity
sp.oo + 15 * sp.pi

oo

Still we get infinity. 

The infinity constant works the way think it does:

In [16]:
# Subtract infinity from infinity
sp.oo - sp.oo

nan

In [17]:
# The imaginary unit
sp.I

I

In [18]:
# Complex arithematic is implemented 
sp.I ** 2

-1

#### **Coomon Functions**
SymPy has loads of cool functions implemented. Let me go a few of the most common ones:

In [19]:
# Exponential function 
sp.exp(x)

exp(x)

In [20]:
# Eulers Identity
sp.exp(sp.pi * sp.I)

-1

In [21]:
# Logarithm function
sp.log(x)

log(x)

In [22]:
# Checking that it is in fact the natural logarithm
sp.exp(sp.log(x))

x

In [23]:
# Trignometric functions 
sp.sin(x) + sp.cos(x) + sp.tan(x)

sin(x) + cos(x) + tan(x)

In [24]:
expr = x**2 + sp.sin(sp.pi * x) + sp.exp(sp.cos(x))
expr

x**2 + exp(cos(x)) + sin(pi*x)